### DISCUSSION TITLE
Help understanding how vector variables are handled in auxkernels

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#22895] [Q&A General] [answered]
Hi all,

I've been trying to learn about auxkernels, with a view to extract velocity gradient components at an arbitrary node ID (not necessarily the node currently under consideration).

From the documentation on the 'couplable' interface, I was hoping that I'd be able to do something along the lines of:
` _vel_grad_x(coupledGradient("velocity", index)) `
where "velocity" is an input vector variable, and 'index' is 0, 1, or 2 for x, y, and z.

Suffice to say this didn't work, and so I began using some cout << statements in my auxkernel to try and work out how to index my velocity vector properly.
Using something along the lines of:
` std::cout << _velocity[currentNodeID] `
printed values looking like this:
` (x,y,z)=(2.18065e-314, 2.47033e-323, 2.18065e-314) `
or for some lines:
` (x,y,z)=(0, 4.82411e+228, 1.90859e+185) `

The e-300 values I thought could be rounding errors and the velocity is zero. However, I can't explain the very large values, and none of the output velocities look close to those in the output file, whose velocity magnitudes, vary between 0 and 8e-2.

My question is, what have I been printing? My hope is that this will clarify how I should set about extracting the velocity gradient components properly.

Thanks!

Below are some snippets from the .h and .C files to clarify how I've defined things in more detail:

In the header file, under 'protected':
` const VectorVariableValue & _velocity; `
` const VariableGradient & _vel_grad_x; `

In the initialiser list in the .C file:
` _velocity(coupledVectorValue("velocity")), `
` _vel_grad_x(coupledGradient("velocity", 0)), `

The cout statement:
` std::cout << _velocity[_current_node->id()] << "\n"; `

Velocity is defined with `family="LAGRANGE_VEC"` in the input file.
I'll also add that my simulation is 2D, so seeing any value in the 'z' component is surprising!

**Comment by GiudGiud** [comment]
Hello

You are reading out of bounds, so there could be any value there.
If velocity is a vector variable, then to retrieve the gradient of it (the whole vector) you will need to save it in a
const VectorVariableGradient & _vel_grad
and use
const [ADVectorVariableGradient](https://mooseframework.inl.gov/docs/doxygen/moose/MooseTypes_8h.html#ac94e300de96c5e84492712644437fcb9) &
to fill it.

Then you will want to access each component with 0, 1, 2 for X Y Z

Guillaume

**Comment by Eilloo** [comment, marked as answer]
Ah, I hadn't seen that VectorVariableGradient is a type!
To make sure I understand, I shouldn't use the 'coupleable' interface to populate it, but instead look to access the ADVectorVariableGradient values directly?

I see now that the nodeID's aren't what I thought and hence are out of bounds garbage when I try to index into _velocity using them... is there a way to convert from a node ID to an index so I can pull out the corresponding variable values?

Thanks for your help!

**Reply by GiudGiud** [reply]
You should use the coupleable interface, but you need to use the right routine to get the vector variable gradient

So that part is tough. The coupleable interface only gets you the local variable value and for some items the neighbor values. It wont get you the general values of the variable everywhere based on the node ID. 

You can use the variable pointer directly (use getVariable to retrieve it) and if you know where to index you can directly access the solution vector (wont work for higher orders very easily)
https://mooseframework.inl.gov/docs/doxygen/moose/classMooseVariableFE.html#acbbf516d918a5de55a3b59832908cf98

or you can use the functor base class of variables. Functors are a base class of variable and help with geometry-based evaluations (like element face, or particular point in space). That hasnt been explored much so far for node IDs, you ll need to do some development
https://mooseframework.inl.gov/docs/doxygen/moose/classMoose_1_1FunctorBase.html

**Reply by GiudGiud** [reply]
> with a view to extract velocity gradient components at an arbitrary node ID (not necessarily the node currently under consideration).

basically this is difficult to do in MOOSE right now. 
Some solutions will either be hacky or involve developments

If you can formulate your problem to always consider the local node, or nodes around the local element, then it will be easier

**Reply by Eilloo** [reply]
Thanks for the info @GiudGiud - I thought I'd say what I've ended up doing, for the sake of completeness:

There's a handy 'find_nodal_neighbors' method in MeshTools, so once you know which node you're interested in, it's easy to get a list of its neighbour nodes using this.
You can then get the dof associated with each velocity component at each of these neighbours using 'dof_number', and changing the 'component' argument. Getting the actual values is achieved in a similar way to the existing 'NearestNodeValue' Auxkernel.

Subtracting these components from those associated with the node of interest essentially gives 'delta_velocity_x' (for, say, the x component), and you can divide this by the difference in x/y/z co-ordinates to get all 4 (2D) or 9 (3D) velocity gradient components.
Averaging each component by summing and dividing by the number of neighbour nodes used to calculate them gives one set of velocity gradient components, associated with the original node of interest.

Now, this isn't very elegant and involves a lot of nested loops, so I suspect things will get very slow very soon when I start looking at more complex meshes... but initial impressions are that it seems to work, in case anyone is interested!

**Reply by GiudGiud** [reply]
Yes that seems workable. Thank you for sharing

**Reply by GiudGiud** [reply]
This won’t work for all variable types obviously, only a subset of nodal variables
